plugins/chat/app/services/services.md

```diff
@@ -0,0 +1,153 @@
+service objects should support both. We can define many ease-of-use methods in
+the `ServiceBase` class to make it much easier to raise and log errors for
+failure states. This leads into the next section.
+
+#### Success and Failure
+
+For most services (excluding pure data fetching ones like ChatChannelFetcher) we
+want to know a result of the service call and any errors that may have been
+encountered that we can show to the user. For this purpose we should have a
+class, called ServiceResult, that encapsulates this. The base service class can
+orchestrate these results, but the gist of it is:
+
+* It has a status indicating whether the service succeeded or failed
+* It has the capacity to store multiple `errors`
+* It has the capacity to store a success `message`
+* It has the capacity to store arbitrary data for successful service calls, e.g.
+  an ActiveRecord model that was created as part of the service call
+
+Then the caller can simply inspect the result of the service method call:
+
+```ruby
+result = ChatChannelArchiveService.create_archive_process(params...)
+if result.failed?
+  render json: failed_json.merge(errors: result.errors), status: result.status
+end
+```
+
+The `result.status` can be derived from the `error_type` defined when adding
+errors above. The most common HTTP status code would be 400, but 403 and 404 are
+relatively common as well. For a success case which serializes the result, it
+would look something more like this:
+
+```ruby
+result = ChatChannelArchiveService.create_archive_process(params...)
+if result.succeeded?
+  render_serialized(ChatChannelArchiveSerializer, result.service_data.chat_channel_archive)
+end
+```
+
+#### Guardian
+
+There are always permissions checks that we will need to do within service
+objects. Sometimes the controller checks some of these in a `before` action and
+other times its up to the service. There is a kind of tension and duplication of
+work effort here, since we may end up checking permissions on something (a
+chat channel for example) multiple times.
+
+I don't think there is an easy solution to this, that doesn't involve some
+contortions around caching Guardian call signatures and their results for a
+specific user inside some sort of wrapper object. For now I think we can leave
+this, unless it's really causing a lot of performance woes.
+
+#### Enqueueing Jobs
+
+This can be done via the base class's `enqueue_job` and `enqueue_job_at` methods,
+rather than calling Jobs.enqueue directly.
+
+#### Logging
+
+As mentioned above, there are two types of logging:
+
+* STDOUT logging using Rails.logger and Discourse.warn_exception
+* UserHistory and StaffActionLogger database-level logs
+
+Both of these can be handled via the base class' `log_message`, `log_user_history`,
+and `log_staff_action` methods.
+
+#### MessageBus
+
+Ideally, any MessageBus interactions should be done via a dedicated MessageBus
+service, such as ChatMessagePublisher.
```

Comment:
All of this shouldn't be part of the service utilities IMO. This is separate concerns.